controller/disk_schedule_controller.go

```diff
@@ -0,0 +1,570 @@
+		return
+	}
+
+	// disk schedule controller prevents data racing by single-thread event handler
+	go wait.Until(dsc.worker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (dsc *DiskScheduleController) worker() {
+	for dsc.processNextWorkItem() {
+	}
+}
+
+func (dsc *DiskScheduleController) processNextWorkItem() bool {
+	key, quit := dsc.queue.Get()
+
+	if quit {
+		return false
+	}
+	defer dsc.queue.Done(key)
+
+	err := dsc.syncDiskSchedule(key.(string))
+	dsc.handleErr(err, key)
+
+	return true
+}
+
+func (dsc *DiskScheduleController) handleErr(err error, key any) {
+	if err == nil {
+		dsc.queue.Forget(key)
+		return
+	}
+
+	log := dsc.logger.WithField("DiskSchedule", key)
+	if dsc.queue.NumRequeues(key) < maxRetries {
+		handleReconcileErrorLogging(log, err, "Failed to sync Longhorn disk schedule")
+		dsc.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	handleReconcileErrorLogging(log, err, "Dropping Longhorn disk schedule out of the queue")
+	dsc.queue.Forget(key)
+}
+
+func (dsc *DiskScheduleController) syncDiskSchedule(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "failed to sync %v", key)
+	}()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != dsc.namespace {
+		// Not ours, don't do anything
+		return nil
+	}
+	return dsc.reconcile(name)
+}
+
+func (dsc *DiskScheduleController) reconcile(diskScheduleName string) (err error) {
+	diskSchedule, err := dsc.ds.GetDiskSchedule(diskScheduleName)
+	if err != nil {
+		if !datastore.ErrorIsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	log := getLoggerForDiskSchedule(dsc.logger, diskSchedule)
+
+	if !diskSchedule.DeletionTimestamp.IsZero() {
+		return dsc.ds.RemoveFinalizerForDiskSchedule(diskSchedule)
+	}
+
+	// disk schedule is controlled only by the owner of the disk
+	if diskSchedule.Spec.NodeID != dsc.controllerID {
+		return nil
+	}
+
+	existingDiskSchedule := diskSchedule.DeepCopy()
+	defer func() {
+		if err != nil {
+			return
+		}
+		if !reflect.DeepEqual(existingDiskSchedule.Spec, diskSchedule.Spec) {
+			if _, err := dsc.ds.UpdateDiskSchedule(diskSchedule); err != nil && apierrors.IsConflict(errors.Cause(err)) {
+				log.WithError(err).Debugf("Requeue %v due to spec conflict", diskScheduleName)
+				dsc.enqueueDiskSchedule(diskSchedule)
+			}
+		}
+		if reflect.DeepEqual(existingDiskSchedule.Status, diskSchedule.Status) {
+			return
+		}
+		if _, updateStatusErr := dsc.ds.UpdateDiskScheduleStatus(diskSchedule); updateStatusErr != nil && datastore.ErrorIsConflict(errors.Cause(updateStatusErr)) {
+			log.WithError(updateStatusErr).Debugf("Requeue %v due to status conflict", diskScheduleName)
+			dsc.enqueueDiskSchedule(diskSchedule)
+		}
+	}()
+
+	if dsc.isDiskScheduleInitialing(diskSchedule) {
+		// blindly record the allocation for existing resource
+		if err := dsc.syncFromAllocatedResources(diskSchedule); err != nil {
```

Comment:
What does Resource mean? If it is space, sugguest renaming it to syncFromAllocatedSpace to make the function easy to understand.